### PR TITLE
use the Resource Manager

### PR DESCRIPTION
**Pull request by marten-seemann**
This is without passing `ReserveMemory` functions to quic-go. We're just using it to open / accept connections and streams.

When reviewing, please pay attention to the hole punching logic.

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -38,14 +40,28 @@ func (c *conn) IsClosed() bool {
...
 func (c *conn) AcceptStream() (mux.MuxedStream, error) {
 	qstr, err := c.sess.AcceptStream(context.Background())
-	return &stream{Stream: qstr}, err
+	if err != nil {
+		return nil, err
+	}
+	if _, err := c.rcmgr.OpenStream(c.remotePeerID, network.DirInbound); err != nil {
+		return nil, err
+	}
+	return &stream{Stream: qstr}, nil
```
Comment (vyzo) [inline conn.go:64]: shouldnt it keep the stream scope?

Comment (vyzo) [reply conn.go:64]: we leak it otherwise

File: listener.go [hunk 2]
```diff
@@ -70,7 +75,13 @@ func (l *listener) Accept() (tpt.CapableConn, error) {
 			sess.CloseWithError(0, err.Error())
 			continue
 		}
-		if l.transport.gater != nil && !(l.transport.gater.InterceptAccept(conn) && l.transport.gater.InterceptSecured(n.DirInbound, conn.remotePeerID, conn)) {
+		connScope, err := l.rcmgr.OpenConnection(network.DirInbound, false)
```
Comment (vyzo) [inline listener.go:78]: where is this passed?

File: transport.go [hunk 3]
```diff
@@ -183,17 +191,25 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
...
 	tlsConf, keyCh := t.identity.ConfigForPeer(p)
-
-	if ok, isClient, _ := n.GetSimultaneousConnect(ctx); ok && !isClient {
-		return t.holePunch(ctx, network, addr, p)
+	if ok, isClient, _ := network.GetSimultaneousConnect(ctx); ok && !isClient {
+		return t.holePunch(ctx, netw, addr, p)
 	}
 
-	pconn, err := t.connManager.Dial(network, addr)
+	scope, err := t.rcmgr.OpenConnection(network.DirOutbound, false)
```
Comment (vyzo) [inline transport.go:198]: and here too, where is the scope passed?

**Review by vyzo — COMMENTED** [review]
we'll need to hook the memory rsvp as well, but ok for now.

File: listener.go (hunk 2 above)
Comment (marten-seemann) [reply listener.go]: Fixed by using the new `transport.CapableConn` interface.

File: transport.go (hunk 3 above)
Comment (marten-seemann) [reply transport.go:198]: Fixed by using the new `transport.CapableConn` interface.

File: conn.go (hunk 1 above)
Comment (marten-seemann) [reply conn.go:65]: The `MuxedStream` doesn't have any scope.
This is probably the wrong layer to do it, this should be done in the swarm.

Comment (marten-seemann) [reply conn.go]: This is now done in the swarm, see https://github.com/libp2p/go-libp2p-swarm/pull/308.

**Review by vyzo — COMMENTED** [review]
we need to Done the scope on closure

File: listener.go [hunk 4]
```diff
@@ -102,20 +105,23 @@ func (l *listener) setupConn(sess quic.Session) (*conn, error) {
...
 	remotePeerID, err := peer.IDFromPublicKey(remotePubKey)
 	if err != nil {
 		return nil, err
 	}
-
 	remoteMultiaddr, err := toQuicMultiaddr(sess.RemoteAddr())
 	if err != nil {
 		return nil, err
 	}
+	connScope, err := l.rcmgr.OpenConnection(network.DirInbound, false)
```
Comment (marten-seemann) [inline listener.go:116]: Should we loop here, or return the error to the caller?

Comment (vyzo) [reply listener.go:116]: let's bail.

**Review by vyzo — APPROVED** [review]

**Pull request merged**